### ISSUE TITLE
compile error at lomse_api_internal_model.cpp(838,62)

### ISSUE THREAD

**Issue by hdhxt11** [lenmus/lomse#253]
compile error:
lomse\src\internal_model\lomse_api_internal_model.cpp(838,62): error C2249: 'lomse::AObject::Private::previous_sibling': no accessible path to inaccessible member declared in virtual base 'lomse::ISiblings'

**Comment by cecilios** [comment]
More details please: Operating system, compiler
Coud you post the log output?

**Comment by hdhxt11** [comment]
> More details please: Operating system, compiler
> Coud you post the log output?

windows 10 , visual studio 2019 version 16.7.6  Debug x64

1>------ Build started: Project: lomse, Configuration: Debug x64 ------
1>lomse_api_internal_model.cpp
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(838,62): error C2249: 'lomse::AObject::Private::previous_sibling': no accessible path to inaccessible member declared in virtual base 'lomse::ISiblings'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(291): message : see declaration of 'lomse::AObject::Private::previous_sibling'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(211): message : see declaration of 'lomse::ISiblings'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(847,58): error C2249: 'lomse::AObject::Private::next_sibling': no accessible path to inaccessible member declared in virtual base 'lomse::ISiblings'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(304): message : see declaration of 'lomse::AObject::Private::next_sibling'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(211): message : see declaration of 'lomse::ISiblings'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(861,1): warning C4819: The file contains a character that cannot be represented in the current code page (936). Save the file in Unicode format to prevent data loss
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(921,71): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(929,71): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(936,67): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(951,67): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(957,67): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(963,67): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(978,67): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(984,67): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(990,67): error C2249: 'lomse::AObject::Private::downcast_content_obj': no accessible path to inaccessible member declared in virtual base 'lomse::IChildren'
1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(216): message : see declaration of 'lomse::AObject::Private::downcast_content_obj'
1>D:\DevPak\Src\lomse\include\lomse_internal_model.h(226): message : see declaration of 'lomse::IChildren'
1>Done building project "lomse.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

**Comment by dmitrio95** [comment]
Looks very similar to this issue:
https://developercommunity.visualstudio.com/content/problem/856080/vc-compilation-error-c2249-accessing-nested-static.html
Similarly to that code example, in this case there is also a virtual inheritance (from `AObject`) and the corresponding `friend` declaration (which is added there by `LOMSE_DECLARE_IM_API_CLASS` macro). I don't see why it shouldn't compile, looks like a compiler bug unless I am missing something.

I wonder though whether virtual inheritance from `AObject` can be avoided here. As far as I can tell, the idea was to be able to define implementation of several generic interfaces for internal objects (`ISiblings` and `IChildren` at the moment) which could be added to concrete objects types independently. However currently every type which is inherited from `IChildren` is also inherited from `ISiblings`, so this independence doesn't seem to be really used at the moment. Given this, would it perhaps make sense to change it to something more like a hierarchical classes structure (something like `AObjectWithSiblings` class which inherits `AObject` and `AObjectWithChildren` which inherits `AObjectWithSiblings`), or is this independent interfaces structure planned to be used later? If not, it looks like virtual inheritance (and multiple inheritance at all) can be avoided here which would possibly simplify the code related to defining API objects.

**Comment by hdhxt11** [comment]
> Looks very similar to this issue:
> https://developercommunity.visualstudio.com/content/problem/856080/vc-compilation-error-c2249-accessing-nested-static.html
> Similarly to that code example, in this case there is also a virtual inheritance (from `AObject`) and the corresponding `friend` declaration (which is added there by `LOMSE_DECLARE_IM_API_CLASS` macro). I don't see why it shouldn't compile, looks like a compiler bug unless I am missing something.
> 
> I wonder though whether virtual inheritance from `AObject` can be avoided here. As far as I can tell, the idea was to be able to define implementation of several generic interfaces for internal objects (`ISiblings` and `IChildren` at the moment) which could be added to concrete objects types independently. However currently every type which is inherited from `IChildren` is also inherited from `ISiblings`, so this independence doesn't seem to be really used at the moment. Given this, would it perhaps make sense to change it to something more like a hierarchical classes structure (something like `AObjectWithSiblings` class which inherits `AObject` and `AObjectWithChildren` which inherits `AObjectWithSiblings`), or is this independent interfaces structure planned to be used later? If not, it looks like virtual inheritance (and multiple inheritance at all) can be avoided here which would possibly simplify the code related to defining API objects.

Thank you very much for this detailed explanation。How do I deal with it?

**Comment by cecilios** [comment]
@dmitrio95   Thank you very much Dmitri.

> looks like a compiler bug unless I am missing something.

I also think this looks like a bug in Microsoft compiler, but I don't know how to bypass it. Moreover, I do not have a Microsoft machine (just maintain another machine with dual boot just because Windows came with it, that I never use) and I don't regularly use Microsoft products, so I have little knowledge for dealing with the peculiarities of that platform. So testing this and finding a bypass will take me some time.

> would it perhaps make sense to change it to something more like a hierarchical classes structure (something like AObjectWithSiblings class which inherits AObject and AObjectWithChildren which inherits AObjectWithSiblings), or is this independent interfaces structure planned to be used later? If not, it looks like virtual inheritance (and multiple inheritance at all) can be avoided here which would possibly simplify the code related to defining API objects.

I like your proposal of defining AObjectWithSiblings and AObjectWithChildren classes to avoid virtual inheritance as, probably, current interfaces IChildren and ISiblings will not be used in any other API classes. But I have to think on it. When I designed the API I tried to avoid that the API classes were simply a wrapper for the internal model classes, as this would prevent changing the internal model when necessary and would expose irrelevant details to the library user. I tried to define closed and independent interfaces for each set of properties (for example ISiblings and IChildren) so that they could be added to those objects that would need them. Trying to follow the Open-Close and Single Responsibility principles, I never considered joining both classes, because although it could be done at this moment, perhaps in the future some object could require inheritance from only one of them. I will study the change you propose. Thank you.

@hdhxt11 Conclusion, I will try to do something to bypass this problem but this will take some time. Sorry! By the way, in the log you submitted there is a warning:

> 1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(861,1): warning C4819: The file contains a character that cannot be represented in the current code page (936). Save the file in Unicode format to prevent data loss

that I do not understand as `lomse_api_internal_model.cpp` is in UTF8 in the repo, and this warning does not appear in AppVeyor log (https://ci.appveyor.com/project/cecilios/lomse-aai68/branch/master). Did you change anything on the sources?

**Comment by hdhxt11** [comment]
> @dmitrio95 Thank you very much Dmitri.
> 
> > looks like a compiler bug unless I am missing something.
> 
> I also think this looks like a bug in Microsoft compiler, but I don't know how to bypass it. Moreover, I do not have a Microsoft machine (just maintain another machine with dual boot just because Windows came with it, that I never use) and I don't regularly use Microsoft products, so I have little knowledge for dealing with the peculiarities of that platform. So testing this and finding a bypass will take me some time.
> 
> > would it perhaps make sense to change it to something more like a hierarchical classes structure (something like AObjectWithSiblings class which inherits AObject and AObjectWithChildren which inherits AObjectWithSiblings), or is this independent interfaces structure planned to be used later? If not, it looks like virtual inheritance (and multiple inheritance at all) can be avoided here which would possibly simplify the code related to defining API objects.
> 
> I like your proposal of defining AObjectWithSiblings and AObjectWithChildren classes to avoid virtual inheritance as, probably, current interfaces IChildren and ISiblings will not be used in any other API classes. But I have to think on it. When I designed the API I tried to avoid that the API classes were simply a wrapper for the internal model classes, as this would prevent changing the internal model when necessary and would expose irrelevant details to the library user. I tried to define closed and independent interfaces for each set of properties (for example ISiblings and IChildren) so that they could be added to those objects that would need them. Trying to follow the Open-Close and Single Responsibility principles, I never considered joining both classes, because although it could be done at this moment, perhaps in the future some object could require inheritance from only one of them. I will study the change you propose. Thank you.
> 
> @hdhxt11 Conclusion, I will try to do something to bypass this problem but this will take some time. Sorry! By the way, in the log you submitted there is a warning:
> 
> > 1>D:\DevPak\Src\lomse\src\internal_model\lomse_api_internal_model.cpp(861,1): warning C4819: The file contains a character that cannot be represented in the current code page (936). Save the file in Unicode format to prevent data loss
> 
> that I do not understand as `lomse_api_internal_model.cpp` is in UTF8 in the repo, and this warning does not appear in AppVeyor log (https://ci.appveyor.com/project/cecilios/lomse-aai68/branch/master). Did you change anything on the sources?

@cecilios Thank you very much Cecilios.I didn't chang anything on the sources. Visual Studio always warns like this. But it doesn't matter.

**Comment by cecilios** [comment]
@hdhxt11   PR #254 should have fix this. Could you please confirm that you can now compile? Thank you!

**Comment by hdhxt11** [comment]
> @hdhxt11 PR #254 should have fix this. Could you please confirm that you can now compile? Thank you!

@cecilios Thank you very much. I will down and comile use VC at once.

Yesterday,I use GCC compiler(gcc (Rev4, Built by MSYS2 project) 10.2.0) at WIN10
found another compile error:(lomse_internal_model.cpp can pass)
D:\DevPak\Src\lomse-master\src\module\lomse_time.cpp: In function 'std::string lomse::to_simple_string(std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long long int, std::ratio<1, 1000000000> > >, bool)':
D:\DevPak\Src\lomse-master\src\module\lomse_time.cpp:77:5: error: 'localtime_r' was not declared in this scope; did you mean 'localtime_s'?
   77 |     localtime_r(&time_sec, &time_tm);
      |     ^~~~~~~~~~~
      |     localtime_s

**Comment by cecilios** [comment]
The GCC error was due, probably, to not having changed the config file for using GCC instead of Microsoft VC. Look at this:
https://github.com/lenmus/lomse/blob/5fd2f47c94f8ce79502851bd8dcc239238792827/src/module/lomse_time.cpp#L69-L79

The CMakeLists.txt script is not prepared to use GCC in Windows. It always assumes Microsoft compiler and thus, the generated config file (lomse_config.h) is not properly configured. So, before compiling it is necessary to change, manually, the lomse_config.h file and change the compiler in following lines:

```
//---------------------------------------------------------------------------------------
// platform and compiler
//---------------------------------------------------------------------------------------
#define LOMSE_PLATFORM_WIN32      1
#define LOMSE_PLATFORM_UNIX       0
#define LOMSE_COMPILER_MSVC       1
#define LOMSE_COMPILER_GCC        0

```
I should add an option to CMakeLists.txt for chosing GCC in Windows.

**Comment by cecilios** [comment]
Solved in PR #254
